packages/utils/src/promise.ts

```diff
@@ -24,3 +24,118 @@ export async function callFnWhenAwait<T>(
+  cutoff = "cutoff-reached",
+  /** atleast one resolved till cutoff so no race required */
+  resolvedatcutoff = "resolved-at-cutoff",
+  /** if none reject/resolve before cutoff but one resolves or all reject before timeout */
+  pretimeout = "pretimeout-return",
+  /** timeout reached as none resolved and some were pending till timeout*/
+  timeout = "timeout-reached",
+
+  // events for the promises for better tracking
+  /** promise resolved */
+  resolved = "resolved",
+  /** promise rejected */
+  rejected = "rejected",
+}
+
+/**
+ * Wait for promises to resolve till cutoff and then race them beyond the cutoff with an overall timeout
+ * @return resolved values or rejections or still pending errors corresponding to input promises
+ */
+export async function racePromisesWithCutoff<T>(
+  promises: Promise<T>[],
+  cutoffMs: number,
+  timeoutMs: number,
+  eventCb: (event: RaceEvent, delayMs: number, index?: number) => void
+): Promise<(Error | T)[]> {
+  // start the cutoff and timeout timers
+  let cutoffObserved = false;
+  const cutoffPromise = new Promise((_resolve, reject) => setTimeout(reject, cutoffMs)).catch((e) => {
+    cutoffObserved = true;
+    throw e;
+  });
+  let timeoutObserved = false;
+  const timeoutPromise = new Promise((_resolve, reject) => setTimeout(reject, timeoutMs)).catch((e) => {
+    timeoutObserved = true;
+    throw e;
+  });
+  const startTime = Date.now();
+
+  // Track promises status and resolved values/rejected errors
+  // Even if the promises reject with the following decoration promises will not throw
+  const promisesStates = [] as PromiseState<T>[];
+  promises.forEach((promise, index) => {
+    promisesStates[index] = {status: PromiseStatus.pending, value: null};
+    promise
+      .then((value) => {
+        eventCb(RaceEvent.resolved, Date.now() - startTime, index);
+        promisesStates[index] = {status: PromiseStatus.resolved, value};
+      })
+      .catch((e: Error) => {
+        eventCb(RaceEvent.rejected, Date.now() - startTime, index);
+        promisesStates[index] = {status: PromiseStatus.rejected, value: e};
+      });
+  });
+
+  // Wait till cutoff time unless all original promises resolve/reject early
+  await Promise.allSettled(promises.map((promise) => Promise.race([promise, cutoffPromise])));
+  if (cutoffObserved) {
+    // If any is resolved, then just simply return as we are post cutoff
+    const anyResolved = promisesStates.reduce(
+      (acc, pmState) => acc || pmState.status === PromiseStatus.resolved,
+      false
+    );
+    if (anyResolved) {
+      eventCb(RaceEvent.resolvedatcutoff, Date.now() - startTime);
+      return mapStatusesToResponses(promisesStates);
+    } else {
+      eventCb(RaceEvent.cutoff, Date.now() - startTime);
+    }
+  } else {
+    eventCb(RaceEvent.precutoff, Date.now() - startTime);
+    return mapStatusesToResponses(promisesStates);
+  }
+
+  // Post deadline resolve with any of the promise or all rejected before timeout
```

Comment:
deadline is the cutoff here, right?

```suggestion
  // Post cutoff resolve with any of the promise or all rejected before timeout
```